### PR TITLE
Fix non-periodic boundary index space

### LINKED ISSUE #615: Incorrect boundaryIndexSpace non-periodic physical boundaries indexes
With physical boundaries, boundaryIndexSpace for direction 1 in a dimension has an extent of 3 not 2. This is incorrect for non-periodic boundaries, and if you iterate over the resulting index space, can can index outside of the valid indexes of the associated array view. A simple reproducer is shown below:

```
#include <Cajita.hpp>
#include <Kokkos_Core.hpp>
​
using MeshDeviceType = Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>;
​
int main( int argc, char* argv[] )
{
    int rank;
    MPI_Init( &argc, &argv );
    Kokkos::initialize( argc, argv );
​
    MPI_Comm_rank( MPI_COMM_WORLD, &rank );
​
    // Create the global mesh and grid. 4 cells, 5 node (non-periodic)
    // in each of 2 directions (i and j)
    std::array<double, 2> low = {-2.0, -2.0}, high = {2.0, 2.0};
    auto global_mesh = Cajita::createUniformGlobalMesh( low, high, 1.0 );
​
    Cajita::DimBlockPartitioner<2> partitioner; // Create Cajita Partitioner
    auto global_grid = Cajita::createGlobalGrid( MPI_COMM_WORLD, global_mesh,
                                                 {false, false}, partitioner );
    // Build the local grid.
    int halo_width = 2;
    auto local_grid = Cajita::createLocalGrid( global_grid, halo_width );
​
    // The resulting local grid should have 6 cells and 7 nodes
   
    // Now create an array layout for that.
    auto node_triple_layout =
            Cajita::createArrayLayout( local_grid, 3, Cajita::Node() );
​
    // Make an array and view from that layout
    auto array = Cajita::createArray<double, MeshDeviceType>(
                     "test array", node_triple_layout );
​
    // Check that the view has 9 entries in each direction
    for (int d = 0; d < 2; d++) {
        auto view = array->view();
        int extent = view.extent(d);
        if (extent != 9) {
            std::cout << "Extent of view in direction " << d << " is " << extent << "\n";
        }
    }
  
    // Now get the boundary index space in each direction.
    for (int i = -1; i <= 1; i++) {
        for (int j = -1; j <= 1; j++) {
            if (i == 0 && j == 0) continue;
​
            std::array<int, 2> dir = {i, j};
            auto boundary_space = local_grid->boundaryIndexSpace(Cajita::Ghost(),
                              Cajita::Node(), dir);
​
            for (int d = 0; d < 2; d++) {
                int bwidth = boundary_space.max(d) - boundary_space.min(d);
                if ((dir[d] != 0) && (bwidth != 2)) {
                    std::cout << "Width in direction "
                              << "{" << dir[0] << ", " << dir[1] << "}"
                              <<  " dim " << d << " is " << bwidth << " instead of 2.\n";
                }
            }
        }
    }
    array = NULL;
​
    Kokkos::finalize();
    MPI_Finalize();
    return 0;
}
```

### PR DESCRIPTION
**Pull request by streeve**
Fixes #615 

1. Add missing tests for `Node`, `Face`, and `Edge`
2. Fix in source

### CODE REVIEW

**Comment by streeve** [comment]
@kwitaechong I added a test to show how this is still failing (as in the original issue)

**Comment by kwitaechong** [comment]
@streeve `for ( long l = 0; l < ghosted_space.extent( 2 ); ++l )` was an issue. It is fixed and pushed toward yours.

**Comment by streeve** [comment]
> `for ( long l = 0; l < ghosted_space.extent( 2 ); ++l )` was an issue. It is fixed and pushed toward yours.

My mistake. I added a more robust set of tests across all cases to be more sure we're not going out of bounds

**Review by streeve — COMMENTED** [review]
A few initial comments

File: cajita/unit_test/tstLocalGrid.hpp [hunk 1]
```diff
@@ -1859,7 +1859,7 @@ void notPeriodicTest3d()
     auto grid_comm = global_grid->comm();
 
     // Create a local mesh.
-    int halo_width = 2;
+    int halo_width = 5;
```
[the remark this answers is not in the document]
Comment (kwitaechong) [reply cajita/unit_test/tstLocalGrid.hpp:1862]: nothing special. Had to use this `5` if you want to test various specified halo width such as 1,3,4

File: cajita/src/Cajita_LocalGrid_impl.hpp [hunk 2]
```diff
@@ -207,7 +207,7 @@ auto LocalGrid<MeshType>::boundaryIndexSpace(
 
     // The boundary index space is just the shared index space for the
```
[the remark this answers is not in the document]
Comment (kwitaechong) [reply cajita/src/Cajita_LocalGrid_impl.hpp:208]: done

File: cajita/unit_test/tstLocalGrid.hpp [hunk 3]
```diff
@@ -3279,21 +4752,24 @@ void notPeriodicTest2d()
     // specified halo width.
     if ( -1 == local_grid->neighborRank( -1, 0 ) )
     {
+        int new_hw = 1;
         boundary_cell_space =
-            local_grid->boundaryIndexSpace( Ghost{}, Cell(), -1, 0, 1 );
-        EXPECT_EQ( boundary_cell_space.min( Dim::I ), halo_width - 1 );
-        EXPECT_EQ( boundary_cell_space.max( Dim::I ), halo_width );
+            local_grid->boundaryIndexSpace( Ghost{}, Cell(), -1, 0, new_hw );
+        EXPECT_EQ( boundary_cell_space.min( Dim::I ),
```
[the remark this answers is not in the document]
Comment (kwitaechong) [reply cajita/unit_test/tstLocalGrid.hpp:4758]: No, I did not touch `Cell` but used `new_hw = 1`.

**Comment by kwitaechong** [comment]
> 

looks good!

**Comment by streeve** [comment]
@kwitaechong can you try to rewrite the boundary tests for Face/Edge the same way I did for Cell/Node?

**Comment by streeve** [comment]
@sslattery @abisner this is ready for review. Will clean up the commits after

File: cajita/unit_test/tstLocalGrid.hpp [hunk 4]
```diff
@@ -27,6 +27,192 @@ using namespace Cajita;
...
+                        const std::array<int, NSD> neighbor,
+                        const int min_lower_shift, const int min_upper_shift,
+                        const int max_lower_shift, const int max_upper_shift,
+                        const std::size_t Dir )
+{
+    for ( std::size_t d = 0; d < NSD; ++d )
+    {
+        int min_upper_shift_dir = 0;
+        int max_lower_shift_dir = 0;
+        if ( Dir != d )
```
Comment (streeve) [inline cajita/unit_test/tstLocalGrid.hpp:119]: Am I missing something here? I would expect edges to end up the same as faces

File: .github/workflows/Nightly.yml [hunk 5]
```diff
@@ -4,6 +4,8 @@ on:
     inputs:
       cmake_args:
         description: 'Extra CMake args'
+      ctest_args:
```
Comment (streeve) [inline .github/workflows/Nightly.yml:7]: Needs to be rebased

File: cajita/unit_test/tstLocalGrid.hpp (hunk 4 above)
Comment (kwitaechong) [reply cajita/unit_test/tstLocalGrid.hpp:119]: `edge` behave differently when `Dir == d`, it behave like a `Cell`. But `face` behave like a `Node` when `Dir == d`

File: .github/workflows/Nightly.yml (hunk 5 above)
Comment (kwitaechong) [reply .github/workflows/Nightly.yml]: done

Comment (streeve) [reply .github/workflows/Nightly.yml]: Need to be careful when you do that - my last commits were overwritten

Comment (kwitaechong) [reply .github/workflows/Nightly.yml]: sorry, anything to be fixed?
